File: projeto-cidades-front/src/app/core/service/state.service.ts
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { HttpClient } from '@angular/common/http';
import { MatSnackBar } from '@angular/material/snack-bar';

import { map, catchError, finalize } from 'rxjs/operators';
import { Observable } from 'rxjs';

import { Endpoints } from "../endpoints.enum";
import { MessageEncapsuling } from '../models/message-encapsuling.model';
import { LoadingService } from './loading.service';
import { StateCounterDto } from '../models/state-counter-dto.model';
import { State } from '../models/state.model';

@Injectable({ providedIn: 'root' })
export class StateService{

    constructor(
        protected router: Router,
        protected http: HttpClient,
        private snackBar: MatSnackBar,
        private loadingService: LoadingService) {

    }
    /**
     * Return the highest and the lowest state ranked by number of cities
     * @returns the MessageEncapsuling containing the highest and lowest
     */
    public getHightestAndLowest(): Observable<MessageEncapsuling<StateCounterDto[]>>{
        this.loadingService.setLoading();
        return this.http.get(Endpoints.HIGHEST_AND_LOWEST_STATE)
        .pipe(map(res => new MessageEncapsuling<StateCounterDto[]>(res)),
        catchError((error: any) => {
            this.snackBar.open((<MessageEncapsuling<StateCounterDto[]>>error.error).message.toString());
            return Observable.throw(error);
        }), finalize(() => this.loadingService.unLoad()));
    }
    /**
     * Return all the states
     * @returns the MessageEncapsuling containing the highest and lowest
     */
    public getAll(): Observable<MessageEncapsuling<State[]>>{
        this.loadingService.setLoading();
        return this.http.get(Endpoints.ALL_STATES)
        .pipe(map(res => new MessageEncapsuling<State[]>(res)),
        catchError((error: any) => {
            this.snackBar.open((<MessageEncapsuling<State[]>>error.error).message.toString());
            return Observable.throw(error);
        }), finalize(() => this.loadingService.unLoad()));
    }
}

File: projeto-cidades-integra-back/src/main/java/com/jamalxvi/projetocidadesintegra/projetocidadesintegra/service/StateService.java
package com.jamalxvi.projetocidadesintegra.projetocidadesintegra.service;

import com.jamalxvi.projetocidadesintegra.projetocidadesintegra.dto.StateCounterDto;
import com.jamalxvi.projetocidadesintegra.projetocidadesintegra.models.MessageEncapsuling;
import com.jamalxvi.projetocidadesintegra.projetocidadesintegra.models.State;

import java.util.List;

public interface StateService {
    /***
     * Find the state that contains the highest number of cities and the state that has the
     * lowest number os cities
     * @return a List containing the hightest populated (in number of cities) state and the lowest
     */
    public MessageEncapsuling<List<StateCounterDto>> hightestAndLowestStatesPopulateOfCities();

    /***
     * Find all the states
     * @return a List containing all the states
     */
    public MessageEncapsuling<List<State>> findAll();
}
